File: 0101-0200/0131.Palindrome_Partitioning.js
//33ms



var partition = function(s) {
    const isPalindrome = (sub) => {
        let left = 0, right = sub.length - 1;
        while (left < right) {
            if (sub[left++] !== sub[right--]) {
                return false;
            }
        }
        return true;
    };
    
    const backtrack = (start, path, partitions) => {
        if (start === s.length) {
            partitions.push([...path]);
            return;
        }
        
        for (let end = start + 1; end <= s.length; end++) {
            const sub = s.substring(start, end);
            if (isPalindrome(sub)) {
                path.push(sub);
                backtrack(end, path, partitions);
                path.pop();
            }
        }
    };
    
    const partitions = [];
    backtrack(0, [], partitions);
    return partitions;
};




//13ms




/**
 * @param {string} s
 * @return {string[][]}
 */

const isPalindrome = (str) =>{
    let left = 0;
    let right = str.length - 1;

    while(left < right){
        if(str[left++] != str[right--]) return false;
    }
    return true;
}

const generate = (
    result, str, len, idx, list
)=>{
    if(idx === len){
        result.push([...list]);
        return;
    }

    let currStr = ""

    for(let i = idx; i < len; i++){
        currStr += str[i];
        
        if(isPalindrome(currStr)){
            list.push(currStr);
            generate(result, str, len, i + 1, list)
            list.pop();
        }
    }
}
const partition = (s)=> {
    const result = [];
    generate(result, s, s.length, 0, []);
    return result;
};

partition("aab");
partition("a");





//11ms




/**
 * @param {string} s
 * @return {string[][]}
 */
function isPalindrome(s, i, j, dp) {
    if (i >= j) return true;
    if (s[i] !== s[j]) return false;
    if (dp[i][j] !== -1) return dp[i][j];

    dp[i][j] = isPalindrome(s, i + 1, j - 1, dp);
    return dp[i][j];
}

var partition = function (s) {
    const n = s.length;
    const res = [];
    const temp = [];
    const dp = Array.from({ length: n+1 }, () => new Array(n+1).fill(-1));

    function solve(idx) {
        if (idx === n) {
            res.push([...temp]);
            return;
        }

        for(let j=idx; j<n; j++) {
            const str = s.slice(idx, j+1);
            if (isPalindrome(s, idx, j, dp)) {
                temp.push(str);
                solve(j+1);
                temp.pop();
            }
        }
    }

    solve(0);
    return res;
};

File: 0101-0200/0131.Palindrome_Partitioning.rs
//21ms



impl Solution {
    fn partition(s: String) -> Vec<Vec<String>> {
        // backtracking 
        // 1. define `ans` to hold `candidates`
        let mut ans = Vec::new();
        // 2. define `candidates` to hold all potential candidates (palindromic substring)
        let mut candidates = Vec::new();
        // 3. start backtrack from index 0
        Solution::backtrack(&s, &mut ans, &mut candidates, 0);
        // 4. return ans
        ans
    }

    fn backtrack(s: &String, ans: &mut Vec<Vec<String>>, candidates: &mut Vec<String>, start: usize) {
        // 1. check if the goal is fulfilled, 
        // i.e. reaching the end of string in this problem
        if start == s.len() {
            // if so, we push `candidates` to ans since we've processed all characters
            ans.push(candidates.clone());
            return;
        }
        // 2. find all potential candidates
        for i in start..s.len() {
            // we want to test all substrings, each substring is a potential candidate
            // e.g. "aab" -> "a", "a", "b", "ab", "aa", "b", "aab"
            let candidate = &s[start .. i + 1];
            // 3. check if the current candidate is palindrome or not 
            // if not, then we cannot push to `candidates`
            if !Solution::is_palindrome(candidate) {
                continue;
            }
            // 4. if so, push it to candidates
            candidates.push(candidate.to_string());
            // 5. backtrack with index + 1
            Solution::backtrack(s, ans, candidates, i + 1);
            // 6. remove the current substring from `candidates`
            candidates.pop();
        }
    }

    fn is_palindrome(s: &str) -> bool {
        s.chars().eq(s.chars().rev())
    }
}




//16ms




impl Solution {
    pub fn partition(s: String) -> Vec<Vec<String>> {
        Self::partition_real(&s)
    }

    fn partition_real(s: &str) -> Vec<Vec<String>> {
        if s.is_empty() {
            return vec![];
        }

        let mut result = vec![];
        let mut stack = vec![];

        Self::partition_recursive(s, &mut result, &mut stack);

        result
    }

    fn partition_recursive<'a>(
        s: &'a str,
        result: &mut Vec<Vec<String>>,
        stack: &mut Vec<&'a str>,
    ) {
        if s.is_empty() {
            result.push(stack.iter().map(|s| s.to_string()).collect());
        }

        for split in 1..=s.len() {
            let (prefix, rest) = s.split_at(split);

            if !Self::is_palindrome(prefix) {
                continue;
            }

            stack.push(prefix);
            Self::partition_recursive(rest, result, stack);
            stack.pop();
        }
    }

    fn is_palindrome(s: &str) -> bool {
        let mut left = 0;
        let mut right = s.len() - 1;

        while left < right {
            if s.as_bytes()[left] != s.as_bytes()[right] {
                return false;
            }

            left += 1;
            right -= 1;
        }

        true
    }
}




//19ms




impl Solution {
    pub fn partition(s: String) -> Vec<Vec<String>> {
        let chars: Vec<char> = s.chars().collect();
        let mut answer = Vec::new();
        Self::helper(
            0,
            &mut Vec::new(),
            &mut answer,
            &chars,
        );
        answer
    }

    fn helper(
        index: usize,
        answer_cache: &mut Vec<String>,
        answer: &mut Vec<Vec<String>>,
        chars: &[char],
    ) {
        if index == chars.len() {
            answer.push(answer_cache.clone());
            return;
        }
        let mut end = index + 1;
        for end in (index + 1)..(chars.len() + 1) {
            if Self::is_palindrome(&chars[index..end]) {
                answer_cache.push(
                    chars[index..end].iter().collect()
                );
                Self::helper(
                    end,
                    answer_cache,
                    answer,
                    chars,
                );
                answer_cache.pop();
            }
        }
    }

    fn is_palindrome(chars: &[char]) -> bool {
        let mid = chars.len() / 2;
        for i in 0..mid {
            if chars[i] != chars[chars.len() - i - 1] {
                return false;
            }
        }
        true
    }
}
